dot/rpc/modules/state_test.go

```diff
@@ -797,89 +799,306 @@ func TestStateModuleGetStorageSize(t *testing.T) {
+					StartBlock: common.Hash{2},
+					EndBlock:   common.Hash{99},
 				},
 			},
 			exp:    []StorageChangeSetResponse{},
-			expErr: errors.New("the start block hash cannot be an empty value"),
+			expErr: chaindb.ErrKeyNotFound,
+		},
+		"start block/end block/error get hash by number": {
+			fields: fields{func(ctrl *gomock.Controller) StorageAPI {
+				mockStorageAPI := NewMockStorageAPI(ctrl)
+				return mockStorageAPI
+			},
+				func(ctrl *gomock.Controller) BlockAPI {
+					mockBlockAPI := NewMockBlockAPI(ctrl)
+					mockBlockAPI.EXPECT().GetBlockByHash(common.Hash{2}).Return(&types.Block{
+						Header: types.Header{
+							Number: 1,
+						},
+					}, nil)
+					mockBlockAPI.EXPECT().GetBlockByHash(common.Hash{3}).Return(&types.Block{
+						Header: types.Header{Number: 2},
+					}, nil)
+					mockBlockAPI.EXPECT().GetHashByNumber(uint(1)).Return(common.Hash{}, blocktree.ErrNumLowerThanRoot)
+					return mockBlockAPI
+				}},
+			args: args{
+				req: &StateStorageQueryRangeRequest{
+					Keys:       []string{"0x010204"},
+					StartBlock: common.Hash{2},
+					EndBlock:   common.Hash{3},
+				},
+			},
+			exp:    []StorageChangeSetResponse{},
+			expErr: blocktree.ErrNumLowerThanRoot,
+		},
+		"start block/end block/error get storage by block hash": {
+			fields: fields{func(ctrl *gomock.Controller) StorageAPI {
+				mockStorageAPI := NewMockStorageAPI(ctrl)
+				mockStorageAPI.EXPECT().GetStorageByBlockHash(&common.Hash{2}, []byte{1, 2, 4}).Return(nil, chaindb.ErrKeyNotFound)
+				return mockStorageAPI
+			},
+				func(ctrl *gomock.Controller) BlockAPI {
+					mockBlockAPI := NewMockBlockAPI(ctrl)
+					mockBlockAPI.EXPECT().GetBlockByHash(common.Hash{2}).Return(&types.Block{
+						Header: types.Header{
+							Number: 1,
+						},
+					}, nil)
+					mockBlockAPI.EXPECT().GetBlockByHash(common.Hash{3}).Return(&types.Block{
+						Header: types.Header{Number: 2},
+					}, nil)
+					mockBlockAPI.EXPECT().GetHashByNumber(uint(1)).Return(common.Hash{2}, nil)
+					return mockBlockAPI
+				}},
+			args: args{
+				req: &StateStorageQueryRangeRequest{
+					Keys:       []string{"0x010204"},
+					StartBlock: common.Hash{2},
+					EndBlock:   common.Hash{3},
+				},
+			},
+			exp:    []StorageChangeSetResponse{},
+			expErr: chaindb.ErrKeyNotFound,
 		},
 	}
-	for _, tt := range tests {
-		t.Run(tt.name, func(t *testing.T) {
+	for name, tt := range tests {
+		tt := tt
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			ctrl := gomock.NewController(t)
 			sm := &StateModule{
-				networkAPI: tt.fields.networkAPI,
-				storageAPI: tt.fields.storageAPI,
-				coreAPI:    tt.fields.coreAPI,
+				storageAPI: tt.fields.storageAPIBuilder(ctrl),
+				blockAPI:   tt.fields.blockAPIBuilder(ctrl),
 			}
 			res := []StorageChangeSetResponse{}
 			err := sm.QueryStorage(tt.args.in0, tt.args.req, &res)
 			if tt.expErr != nil {
-				assert.EqualError(t, err, tt.expErr.Error())
+				assert.ErrorContains(t, err, tt.expErr.Error())
```

Comment:
nit Why is this? Is the error string containing random stuff? Maybe use a regex instead to assert the string matches the expected regular expression?

Reply:
I'd like to leave this as ErrorContains instead of using regular expression for simplicity.

Reply:
It looks rather confusing to compare against `tt.expErr.Error()` then, maybe then change tt.expErr to be `expectedErrStringContained string`?

Although you can use `assert.Regexp(t, tt.errRegexp, err.Error())` which imo is simpler anyway.

Also is it possible to assert the wrapped error with `assert.ErrorIs(t, err, tt.expErr)` (where `tt.expErr` is a sentinel error)?

Reply:
Updated to use regular expression